File: js/engine/gear.js
/**
 * Virtual Drivetrain — realistic gear system for indoor cycling.
 *
 * Models a real cassette with chainring and rear cogs.
 * Each gear has a physical ratio (front teeth / rear teeth) that
 * determines how much resistance offset is sent to the trainer.
 *
 * On a real bike, harder gears at the same cadence produce higher
 * wheel speed and therefore more aerodynamic drag. On a smart trainer
 * in SIM mode, we simulate this by adding a grade offset:
 *
 *   effectiveGrade = routeSlope × trainerDifficulty + gearGradeOffset
 *
 * This ensures gear changes affect resistance on flats, climbs, and descents.
 *
 * Power = Torque × AngularVelocity — at constant power:
 *   - Harder gear → more torque, lower cadence
 *   - Easier gear → less torque, higher cadence
 *   - The trainer resistance reflects this via the grade offset
 */

// ── Default cassette: 50T front × 11-speed rear ─────────────────
// Ordered easiest (largest cog) to hardest (smallest cog).
const DEFAULT_REAR_COGS = [32, 28, 25, 22, 20, 18, 16, 14, 13, 12, 11];
const DEFAULT_CHAINRING = 50;
const DEFAULT_WHEEL_CIRC = 2.105; // meters — 700c × 25mm tire

const DEFAULT_ROLLER_MIN_GRADE = 1;
const DEFAULT_ROLLER_MAX_GRADE = 22;
const DEFAULT_VIRTUAL_GEAR_COUNT = 22;

export class GearSystem {
    /**
     * @param {Object} [config]
     * @param {number[]} [config.rearCogs] — tooth counts, easiest→hardest
     * @param {number}   [config.chainring] — front chainring teeth
     * @param {number}   [config.neutralGear] — index of neutral gear (null = middle)
     * @param {number}   [config.virtualGearCount] — number of virtual gears
     * @param {number}   [config.rollerMinGrade] — lowest gear grade offset (%)
     * @param {number}   [config.rollerMaxGrade] — highest gear grade offset (%)
     * @param {number}   [config.wheelCircumference] — wheel circ in meters
     * @param {number}   [config.debounceMs] — minimum ms between shifts
     * @param {number}   [config.smoothing] — EMA alpha for smooth transitions (0–1)
     */
    constructor(config = {}) {
        this.rearCogs     = config.rearCogs || [...DEFAULT_REAR_COGS];
        this.chainring    = config.chainring || DEFAULT_CHAINRING;
        this.wheelCirc    = config.wheelCircumference || DEFAULT_WHEEL_CIRC;
        this.debounceMs   = config.debounceMs ?? 200;
        this.smoothing    = config.smoothing ?? 0.3;
        this.rollerMinGrade = normalizeGrade(config.rollerMinGrade, DEFAULT_ROLLER_MIN_GRADE);
        this.rollerMaxGrade = normalizeGrade(config.rollerMaxGrade, DEFAULT_ROLLER_MAX_GRADE);
        if (this.rollerMinGrade > this.rollerMaxGrade) {
            [this.rollerMinGrade, this.rollerMaxGrade] = [this.rollerMaxGrade, this.rollerMinGrade];
        }

        this.gearCount    = normalizeGearCount(config.virtualGearCount, DEFAULT_VIRTUAL_GEAR_COUNT);
        this.gearMin      = 0;
        this.gearMax      = this.gearCount - 1;

        // Neutral gear — defaults to middle of cassette
        const defaultNeutral = Math.floor(this.gearCount / 2);
        this.neutralGear  = config.neutralGear ?? defaultNeutral;
        this.neutralGear  = Math.max(this.gearMin, Math.min(this.neutralGear, this.gearMax));

        // Precompute virtual ratios across the physical cassette range. Gear 0
        // maps to rollerMinGrade; the final gear maps to rollerMaxGrade.
        const easiestRatio = this.chainring / this.rearCogs[0];
        const hardestRatio = this.chainring / this.rearCogs[this.rearCogs.length - 1];
        this._ratios = Array.from({ length: this.gearCount }, (_, index) => {
            if (this.gearMax <= this.gearMin) return easiestRatio;
            const t = index / this.gearMax;
            return easiestRatio + ((hardestRatio - easiestRatio) * t);
        });

        // State
        this.currentGear     = this.gearMin;
        this._lastShiftTime  = 0;
        this._targetOffset   = this._computeOffset(this.currentGear);
        this._smoothOffset   = this._targetOffset;
    }

    // ──────────────────────────────────────────────────────────────
    //  SHIFTING
    // ──────────────────────────────────────────────────────────────

    /**
     * Shift to a harder gear (higher ratio, smaller rear cog).
     */
    shiftUp() {
        const now = performance.now();
        if ((now - this._lastShiftTime) < this.debounceMs) return;
        if (this.currentGear >= this.gearMax) return;

        const old = this.currentGear;
        this.currentGear += 1;
        this._targetOffset = this._computeOffset(this.currentGear);
        this._lastShiftTime = now;
        console.log(`[GEAR] ${old}→${this.currentGear} (HARDER) | ratio: ${this.getRatio().toFixed(2)} | roller: ${this._targetOffset.toFixed(1)}`);
    }

    /**
     * Shift to an easier gear (lower ratio, larger rear cog).
     */
    shiftDown() {
        const now = performance.now();
        if ((now - this._lastShiftTime) < this.debounceMs) return;
        if (this.currentGear <= this.gearMin) return;

        const old = this.currentGear;
        this.currentGear -= 1;
        this._targetOffset = this._computeOffset(this.currentGear);
        this._lastShiftTime = now;
        console.log(`[GEAR] ${old}→${this.currentGear} (EASIER) | ratio: ${this.getRatio().toFixed(2)} | roller: ${this._targetOffset.toFixed(1)}`);
    }

    // ──────────────────────────────────────────────────────────────
    //  GRADE OFFSET COMPUTATION
    // ──────────────────────────────────────────────────────────────

    /**
     * Compute the grade offset for a given gear index.
     * Offset is linearly mapped across the configured roller range.
     */
    _computeOffset(gearIndex) {
        if (this.gearMax <= this.gearMin) return this.rollerMinGrade;
        const t = (gearIndex - this.gearMin) / (this.gearMax - this.gearMin);
        return this.rollerMinGrade + ((this.rollerMaxGrade - this.rollerMinGrade) * t);
    }

    configure(config = {}) {
        const previousMax = Math.max(this.gearMax, 1);
        const currentPosition = this.currentGear / previousMax;

        if (config.virtualGearCount !== undefined) {
            this.gearCount = normalizeGearCount(config.virtualGearCount, this.gearCount);
            this.gearMax = this.gearCount - 1;
            this.neutralGear = Math.max(this.gearMin, Math.min(Math.floor(this.gearCount / 2), this.gearMax));

            const easiestRatio = this.chainring / this.rearCogs[0];
            const hardestRatio = this.chainring / this.rearCogs[this.rearCogs.length - 1];
            this._ratios = Array.from({ length: this.gearCount }, (_, index) => {
                if (this.gearMax <= this.gearMin) return easiestRatio;
                const t = index / this.gearMax;
                return easiestRatio + ((hardestRatio - easiestRatio) * t);
            });
            this.currentGear = Math.max(this.gearMin, Math.min(Math.round(currentPosition * this.gearMax), this.gearMax));
        }

        if (config.rollerMinGrade !== undefined) this.rollerMinGrade = normalizeGrade(config.rollerMinGrade, this.rollerMinGrade);
        if (config.rollerMaxGrade !== undefined) this.rollerMaxGrade = normalizeGrade(config.rollerMaxGrade, this.rollerMaxGrade);
        if (this.rollerMinGrade > this.rollerMaxGrade) {
            [this.rollerMinGrade, this.rollerMaxGrade] = [this.rollerMaxGrade, this.rollerMinGrade];
        }

        this._targetOffset = this._computeOffset(this.currentGear);
        this._smoothOffset = this._targetOffset;
    }

    /**
     * Must be called each tick to smooth the transition.
     * @param {number} dt — time step in seconds
     */
    update(dt) {
        if (dt <= 0) return;
        // EMA smoothing toward target offset
        this._smoothOffset += this.smoothing * (this._targetOffset - this._smoothOffset);
        // Snap when close enough
        if (Math.abs(this._targetOffset - this._smoothOffset) < 0.01) {
            this._smoothOffset = this._targetOffset;
        }
    }

    // ──────────────────────────────────────────────────────────────
    //  ACCESSORS
    // ──────────────────────────────────────────────────────────────

    /**
     * Get the current smoothed grade offset (%) to add to trainer grade.
     * Positive = harder, negative = easier.
     */
    getGradeOffset() {
        return Math.round(this._smoothOffset * 100) / 100;
    }

    /** Current gear ratio (chainring / rear cog). */
    getRatio() {
        return this._ratios[this.currentGear];
    }

    /** Current gear index (0 = easiest). */
    getGear() { return this.currentGear; }

    /** Display gear number (1-indexed for UI). */
    getDisplayGear() { return this.currentGear + 1; }

    /** Total number of gears. */
    getGearCount() { return this.gearCount; }

    /** Current equivalent rear cog tooth count. */
    getRearCog() { return Math.round(this.chainring / this.getRatio()); }

    /** Front chainring tooth count. */
    getChainring() { return this.chainring; }

    /**
     * Compute virtual wheel speed from cadence and gear ratio.
     * @param {number} cadenceRPM
     * @returns {number} speed in m/s
     */
    computeWheelSpeed(cadenceRPM) {
        if (cadenceRPM <= 0) return 0;
        return (cadenceRPM / 60) * this.getRatio() * this.wheelCirc;
    }

    /**
     * Reset to neutral gear.
     */
    reset() {
        this.currentGear = this.gearMin;
        this._lastShiftTime = 0;
        this._targetOffset = this._computeOffset(this.currentGear);
        this._smoothOffset = this._targetOffset;
    }
}

function normalizeGrade(value, fallback) {
    const parsed = Number(value);
    if (!Number.isFinite(parsed)) return fallback;
    return Math.max(1, Math.min(40, parsed));
}

function normalizeGearCount(value, fallback) {
    const parsed = Math.round(Number(value));
    if (!Number.isFinite(parsed)) return fallback;
    return Math.max(2, Math.min(40, parsed));
}

File: UNCHAINED_PROJECT/unchained_project/ride/gear.py
"""Virtual gear system — scales trainer gradient feel instead of adding a tiny fixed offset."""

from __future__ import annotations

import time


class GearSystem:
    """Virtual gear system with smooth resistance transitions."""

    def __init__(
        self,
        count=22,
        neutral=11,
        step_grade=0.5,
        max_difficulty_scale=None,
        debounce_ms=200,
        smoothing=0.3,
        min_difficulty_scale=0.15,
        downhill_scale=0.5,
        roller_min_grade=1.0,
        roller_max_grade=22.0,
    ):
        self.gear_min = 0
        self.gear_count = max(2, min(40, int(count)))
        self.gear_max = self.gear_count - 1
        self.gear_neutral = max(self.gear_min, min(neutral, self.gear_max))
        self.step_grade = step_grade
        self.debounce_ms = debounce_ms
        self.smoothing_factor = smoothing
        self.min_difficulty_scale = max(0.0, min(float(min_difficulty_scale), 1.0))
        if max_difficulty_scale is None:
            self.max_difficulty_scale = max(1.0, 1.0 + max(0.0, float(step_grade)))
        else:
            self.max_difficulty_scale = max(1.0, float(max_difficulty_scale))
        self.downhill_scale = max(0.0, min(float(downhill_scale), 1.0))
        self.roller_min_grade = max(1.0, min(40.0, float(roller_min_grade)))
        self.roller_max_grade = max(1.0, min(40.0, float(roller_max_grade)))
        if self.roller_min_grade > self.roller_max_grade:
            self.roller_min_grade, self.roller_max_grade = self.roller_max_grade, self.roller_min_grade

        self.current_gear = self.gear_min
        self._last_shift_time = 0.0
        self._target_offset = self._compute_offset_for_gear(self.current_gear)
        self._smooth_offset = self._target_offset

    def _compute_scale_for_gear(self, gear):
        if gear <= self.gear_neutral:
            lower_span = max(1, self.gear_neutral - self.gear_min)
            ratio = (gear - self.gear_min) / lower_span
            return self.min_difficulty_scale + ((1.0 - self.min_difficulty_scale) * ratio)

        upper_span = max(1, self.gear_max - self.gear_neutral)
        ratio = (gear - self.gear_neutral) / upper_span
        return 1.0 + ((self.max_difficulty_scale - 1.0) * ratio)

    def _compute_offset_for_gear(self, gear):
        if self.gear_max <= self.gear_min:
            return self.roller_min_grade
        ratio = (gear - self.gear_min) / (self.gear_max - self.gear_min)
        return self.roller_min_grade + ((self.roller_max_grade - self.roller_min_grade) * ratio)

    def _update_smooth_offset(self):
        self._smooth_offset += self.smoothing_factor * (self._target_offset - self._smooth_offset)
        if abs(self._target_offset - self._smooth_offset) < 0.01:
            self._smooth_offset = self._target_offset
        return self._smooth_offset

    def shift_up(self):
        """Shift to a harder gear (more resistance)."""
        now = time.time()
        if (now - self._last_shift_time) < (self.debounce_ms / 1000.0):
            print(f"  [GEAR] Shift UP blocked (debounce {self.debounce_ms}ms)")
            return
        if self.current_gear < self.gear_max:
            old = self.current_gear
            self.current_gear += 1
            self._target_offset = self._compute_offset_for_gear(self.current_gear)
            self._last_shift_time = now
            print(f"  [GEAR] {old} → {self.current_gear} (UP) | roller: {self._target_offset:.2f}")
        else:
            print(f"  [GEAR] Already at max gear ({self.gear_max})")

    def shift_down(self):
        """Shift to an easier gear (less resistance)."""
        now = time.time()
        if (now - self._last_shift_time) < (self.debounce_ms / 1000.0):
            print(f"  [GEAR] Shift DOWN blocked (debounce {self.debounce_ms}ms)")
            return
        if self.current_gear > self.gear_min:
            old = self.current_gear
            self.current_gear -= 1
            self._target_offset = self._compute_offset_for_gear(self.current_gear)
            self._last_shift_time = now
            print(f"  [GEAR] {old} → {self.current_gear} (DOWN) | roller: {self._target_offset:.2f}")
        else:
            print(f"  [GEAR] Already at min gear ({self.gear_min})")

    def get_resistance_offset(self, base_slope=0.0):
        """Return the difference between route slope and trainer slope for the current gear."""
        return self._update_smooth_offset()

    def get_gear(self):
        return self.current_gear

    def get_display_gear(self):
        return self.current_gear + 1

    def get_target_offset(self, base_slope=0.0):
        return self._target_offset
